README.md
# rSchedule

[![NPM version](https://flat.badgen.net/npm/v/@rschedule/core)](https://www.npmjs.com/package/@rschedule/core) [![Size when minified & gzipped](https://flat.badgen.net/bundlephobia/minzip/@rschedule/core)](https://bundlephobia.com/result?p=@rschedule/core) [![Actively maintained](https://flat.badgen.net/badge/Maintenance%20level/Actively%20Maintained/green)](https://gist.github.com/cheerfulstoic/d107229326a01ff0f333a1d3476e068d)

### [View docs](#docs)

_The canonical version of this repo is hosted on [Gitlab](https://gitlab.com/john.carroll.p/rschedule), but there is also a [Github mirror](https://github.com/jorroll/rschedule). Issues and merge requests should be created in the [Gitlab repo](https://gitlab.com/john.carroll.p/rschedule)._

A javascript library, written in typescript, for working with recurring dates. The library is "date agnostic" and usable with `Date`, [Moment](https://momentjs.com), [luxon](https://moment.github.io/luxon/), [dayjs](https://day.js.org/), or [js-joda](https://github.com/js-joda/js-joda) objects. If your chosen date library supports time zones, rSchedule supports time zones. All objects in rSchedule are immutable. rSchedule supports creating schedules with durations. rSchedule is modular, tree-shakable, and extensible. It supports JSON and [ICAL](https://tools.ietf.org/html/rfc5545) serialization as well as custom recurrence rules.

```bash
# To install both the main package and the `StandardDateAdapter` for standard javascript dates */

yarn add @rschedule/core @rschedule/standard-date-adapter

# or

npm install @rschedule/core @rschedule/standard-date-adapter

# Current DateAdapter packages

@rschedule/standard-date-adapter
@rschedule/moment-date-adapter
@rschedule/moment-tz-date-adapter
@rschedule/luxon-date-adapter
@rschedule/dayjs-date-adapter
@rschedule/joda-date-adapter
```

[See the docs for setup instructions](#docs)

### Usage ([online demo](https://codesandbox.io/s/rschedule-starter-pxezu?fontsize=14&module=/src/index.ts))

You can play around with a demo of rSchedule on [codesandbox here](https://codesandbox.io/s/rschedule-starter-pxezu?fontsize=14&module=/src/index.ts).

#### Iterate using standard javascript syntax

```typescript
// showing usage with the StandardDateAdapter

const rule = new Rule({
  frequency: 'YEARLY',
  byMonthOfYear: [2, 6],
  byDayOfWeek: ['SU', ['MO', 3]],
  start: new Date(2010, 1, 7),
});

for (const { date } of rule.occurrences({ take: 10 })) {
  date.toISOString();

  // do stuff...
}
```

#### Get the first 5 occurrences after a starting date

```typescript
rule
  .occurrences({
    start: new Date(2010, 5, 7),
    take: 5,
  })
  .toArray()
  .map(date => date.toISOString());
```

#### Add another rule, subtract specific dates, filter to only return unique dates, and query to see if the result occurs on a Monday before `2013/11/15`.

```typescript
const secondRule = new Rule({
  start: new Date(2011, 1, 7),
  frequency: 'DAILY',
  byDayOfWeek: ['MO'],
});

const dates = new Dates({
  dates: [new Date(2010, 10, 15), new Date(2010, 11, 3)],
});

rule
  .pipe(add(secondRule), subtract(dates), unique())
  .occursOn({ weekday: 'MO', before: new Date(2013, 10, 15) }); // true
```

## Docs

- [Version 1.x docs (current)](https://dynodocs.dev/d/gitlab/john.carroll.p/rschedule/branch/v1/docs/1.%20Introduction.md)
- [Version 0.12 docs](https://gitlab.com/john.carroll.p/rschedule/tree/v0.12/docs)
- [Version 0.11 docs](https://gitlab.com/john.carroll.p/rschedule/tree/v0.11/docs)
- [Version 0.10 docs](https://gitlab.com/john.carroll.p/rschedule/tree/f46bf244370dd476633b944e424096a6ae629305/docs)
- [Version 0.9 docs](https://gitlab.com/john.carroll.p/rschedule/tree/a80b576c981570710def8f83575a4932b12f8f34/docs)

[Master branch docs (unreleased)](https://dynodocs.dev/d/gitlab/john.carroll.p/rschedule/branch/master/docs/1.%20Introduction.md)

## Known Limitations

- `@rschedule/ical-tools`
  - No [`BYWEEKNO`](https://gitlab.com/john.carroll.p/rschedule/issues/2), [`BYYEARDAY`](https://gitlab.com/john.carroll.p/rschedule/issues/3), or [`BYSETPOS`](https://gitlab.com/john.carroll.p/rschedule/issues/4) rule support.
  - `VEVENT` supports `RRULE`, `EXRULE`, `RDATE`, `EXDATE`, `DTSTART`, `DTEND` and `DURATION` properties. Other properties are not supported.
  - No `VCALENDAR` iCal support.

## Roadmap

- [x] Explore rearranging library exports and build to better support tree shaking and reduce minimum bundle size.
- [ ] Natural language package for converting rSchedule objects into human readable strings
  - [ ] Internationalization of human readable strings
- [ ] Create `subtractDuration` operator

## Typescript Version

| rSchedule version | Typescript version |
| ----------------- | ------------------ |
| 1.1.3             | 3.5                |
| 1.2.0             | 3.9                |

## About

See the [contributors list](https://gitlab.com/john.carroll.p/rschedule/-/graphs/master) for an up-to-date record of contributions to this repo. The initial version of rSchedule was created by [John Carroll](https://gitlab.com/john.carroll.p). Most of the RRULE tests were taken from the excellent [rrule.js](https://github.com/jakubroztocil/rrule) library (which were, themselves, taken from a python library, I believe).

This library was built using the [typescript starter repo](https://github.com/bitjson/typescript-starter). The implementation strategy drew inspiration from the [Angular Material2](https://github.com/angular/material2) Date Picker component (which makes use of date adapters to support different javascript date libraries), as well as [rrulejs](https://github.com/jakubroztocil/rrule) and [ice_cube](https://github.com/seejohnrun/ice_cube).

_This project is not affiliated with any of these projects._

## Similar libraries

- [rrulejs](https://github.com/jakubroztocil/rrule)
  - Supports time zones via luxon and supports iCal. `rrulejs` is older and more mature than rSchedule and I used it before making rSchedule.
  - For most projects, rrulejs will probably do everything you need and you may feel more comfortable using something older and with a larger install base. Another reason you might want to choose rrule would be for it's NLP, internationalization support, or support for `BYWEEKNO`, `BYYEARDAY`, and `BYSETPOS` ICal rules. By comparison, rSchedule has better timezone support, support for different date libraries, duration support, custom rule support, a smaller minimum bundle size, and complex calendar support. See the docs of both projects to learn more.
- [laterjs](https://github.com/bunkat/later) (currently unmaintained)
  - Simpler API. Not ICAL compatible. Has support for chron jobs.
- [dayspan](https://github.com/ClickerMonkey/dayspan)
  - Appears to be a pretty full featured recurring dates library (like rSchedule or rrulejs), but I don't know much about it. Interestingly, while rrulejs and rSchedule have somewhat similar APIs (borrowing heavily from the ICAL spec), dayspan's API seems to be very different and somewhat unique in places.


CHANGELOG.md
# Change Log

This repo attempts to follow [semantic versioning](https://semver.org/).

## Unreleased

- None

## 1.5.0 FEATURE (2023/2/3)

### Feature

- `@rschedule/core` Adds a `metadata` property to `DateAdapterBase` that can be used to store metadata related to a specific date adapter. Currently this property is just used by the ical-tools package to remember if a given DateAdapter should be serialized as a "date" or "date-time" value. Package consumers can also use this property to store data if they wish.

### Fixes

- `@rschedule/ical-tools` fixes a bug which prevented the proper parsing and serializing of ical "date" values (e.g. `"DTSTART;VALUE=DATE:20210909"`).

## 1.4.0 FEATURE (2022/6/9)

### Feature

- `@rschedule/ical-tools` Add `VEvent.fromJCal()` static method. See [#70](https://gitlab.com/john.carroll.p/rschedule/-/issues/70) for additional information. I don't expect many people to benefit from this.

## 1.3.1 FIX (2021/11/9)

### Fixes

- `@rschedule/luxon-date-adapter` fix typo in `@rschedule/luxon-date-adapter/v1/setup` and `@rschedule/luxon-date-adapter/v2/setup` package.json files

## 1.3.0 FEATURE (2021/10/3)

### Feature

- `@rschedule/luxon-date-adapter` support both luxon `v1.x` and `v2.x`. If you want to use rSchedule with luxon v2, you can do so with the new `v2` entrypoint in the `@rschedule/luxon-date-adapter` package.
  - e.g. `import "@rschedule/luxon-date-adapter/v2/setup"`

## 1.2.4 FIX (2021/1/12)

## Fixes

- Ensure `VEvent#set('start', )` also updates rrules/exrules (issue [#56](https://gitlab.com/john.carroll.p/rschedule/-/issues/56)).

## 1.2.3 FIX (2020/11/19)

## Fixes

- Standardize toISOString() implementation. Previously, different date adapters (apparently) were returning slightly different formatting (e.g. `joda-date-adapter`) ([!62](https://gitlab.com/john.carroll.p/rschedule/-/merge_requests/62)).
- Fix `DatejsDateAdapter#set('timezone', value)`. Previously, attempting to set a new timezone value for a `DayjsDateAdapter` was ignored ([!62](https://gitlab.com/john.carroll.p/rschedule/-/merge_requests/62)).

## 1.2.2 FIX (2020/11/18)

### Fixes

- Do not throw error when reregistering a JSON serializer (issue [#53](https://gitlab.com/john.carroll.p/rschedule/-/issues/53))

## 1.2.1 FIX (2020/9/25)

### Fixes

- Explicitly define the return type of functions and methods (issue [#47](https://gitlab.com/john.carroll.p/rschedule/issues/47))

## 1.2.0 FEATURE (2020/6/25)

### Feature

- Update typescript to version 3.9.x

### Fixes

- Check presence of dayjs "UTC" plugin at run time rather than at import time ([#44](https://gitlab.com/john.carroll.p/rschedule/issues/44)).

## 1.1.3 FEATURE (2020/6/16)

### Feature

- Add `DayjsDateAdapter` package (#43).

## 1.0.5 FIX (2020/6/16)

### Fixes

- Fixed the handling of the `skipToDate` OccurrenceIterator `next()` arg ([!54](https://gitlab.com/john.carroll.p/rschedule/-/merge_requests/54))

## 1.0.4 FIX (2020/6/1)

### Fixes

- Fixed an issue when using IntersectionOperator with a RecurrencePipe ([!53](https://gitlab.com/john.carroll.p/rschedule/-/merge_requests/53))

## 1.0.3 FIX (2019/12/16)

### Fixes

- Properly serialize a VEvent containing a rule with an interval ([#39](https://gitlab.com/john.carroll.p/rschedule/issues/39))
- Remove `byMillisecondOfSecond` from VEvent rule options.

## 1.0.2 FIX (2019/12/1)

- Same as `1.0.1` but fixes a build issue with that version.

## 1.0.1 FIX (2019/12/1)

### Fixes

- Added all side effects declarations to each module's main `package.json` file as an attempted fix for [issue #37](https://gitlab.com/john.carroll.p/rschedule/issues/37). Previously each entry point handled it's own side effects declarations.

## 1.0.0 (2019/11/14)

- This marks the first non-beta release of rSchedule. There is no functional difference between this version and version `0.12.2`.

## 0.12.2 FIX (2019/11/13)

### Fixes

- Clarify the effect of the `timezone` argument in the docs.
- Fixed bug that could occur if a `RecurrenceRulesIterator` was passed run options with a `start`/`end` time in a different time zone from the `RecurrenceRulesIterator's` rule start time ([#35](https://gitlab.com/john.carroll.p/rschedule/issues/35)).
- Fixed a bug that could result in generated date adapters not having the proper `generators` value.

## 0.12.1 FIX (2019/9/26)

### Fixes

- Fixed module build targets so that `UMD` targets `es5` and `es2015` targets `es2015` (#34)

## 0.12.0 BREAKING (2019/9/15)

This is a large breaking change to the library that simplifies the API and increases the modularity / extensibility of the code. You should check out the updated docs to understand all of the changes. The new API is similar to the old API, but better.

#### API improvements

Through some dark, typescript sorcery, rSchedule no longer needs to export generic objects to adapt the library's typing for different date adapters. Now you can simply import your date adapter of choice once, and the types of rSchedule's objects will automatically be updated. Where before you might have:

```ts
const rule = new Rule<typeof MomentDateAdapter>();

const dates: Moment[] = rule
  .occurrences()
  .toArray()
  .map(({ date }) => date);

// OR

const rule = new Rule<typeof StandardDateAdapter>();

const dates: Date[] = rule
  .occurrences()
  .toArray()
  .map(({ date }) => date);
```

now you simply have

```ts
const rule = new Rule();

const dates: Moment[] = rule
  .occurrences()
  .toArray()
  .map(({ date }) => date);

// OR, when using the StandardDateAdapter

const rule = new Rule();

const dates: Date[] = rule
  .occurrences()
  .toArray()
  .map(({ date }) => date);
```

it's magical! End users need never touch a date adapter.

#### Modularity changes


tests/ical.test.ts
import {
  parseByDay,
  parseByHour,
  parseByMinute,
  parseByMonth,
  parseByMonthDay,
  parseBySecond,
  parseCount,
  parseDatetime,
  parseDTStart,
  parseFrequency,
  parseICalStrings,
  parseInterval,
  parseUntil,
  parseWkst,
  ruleOptionsToIcalString,
} from '@rschedule/rschedule';
import { StandardDateAdapter } from '@rschedule/standard-date-adapter';
import { datetime, test } from './utilities';

describe('parseDatetime()', () => {
  describe('VALID', () => {
    test('19990102T030405', time => {
      expect(parseDatetime(time)).toEqual({
        datetimes: [[1999, 1, 2, 3, 4, 5]],
        timezone: undefined,
        raw: time,
      });
    });

    test('20000307T132405Z', time => {
      expect(parseDatetime(time)).toEqual({
        datetimes: [[2000, 3, 7, 13, 24, 5]],
        timezone: 'UTC',
        raw: time,
      });
    });

    test('TZID=America/New_York:19970902T090000', time => {
      expect(parseDatetime(time)).toEqual({
        datetimes: [[1997, 9, 2, 9, 0, 0]],
        timezone: 'America/New_York',
        raw: time,
      });
    });
  });

  describe('INVALID', () => {
    test('20000307132405', time => {
      expect(() => parseDatetime(time)).toThrowError(
        `Invalid ICAL date/time string "${time}"`,
      );
    });
    test('TZID=America/New_York:19970902T09000', time => {
      expect(() => parseDatetime(time)).toThrowError(
        `Invalid ICAL date/time string "${time}"`,
      );
    });
    test('TZIDAmerica/New_York:19970902T090000', time => {
      expect(() => parseDatetime(time)).toThrowError(
        `Invalid ICAL date/time string "${time}"`,
      );
    });
  });
});

describe('parseDTStart()', () => {
  describe('VALID', () => {
    test('DTSTART:19970902T090000', time => {
      expect(parseDTStart(time)).toEqual({
        datetimes: [[1997, 9, 2, 9, 0, 0]],
        timezone: undefined,
        raw: '19970902T090000',
      });
    });

    test('DTSTART:19970902T090000Z', time => {
      expect(parseDTStart(time)).toEqual({
        datetimes: [[1997, 9, 2, 9, 0, 0]],
        timezone: 'UTC',
        raw: '19970902T090000Z',
      });
    });

    test('DTSTART;TZID=America/New_York:19970902T090000', time => {
      expect(parseDTStart(time)).toEqual({
        datetimes: [[1997, 9, 2, 9, 0, 0]],
        timezone: 'America/New_York',
        raw: 'TZID=America/New_York:19970902T090000',
      });
    });
  });

  describe('INVALID', () => {
    test('DTSTART;TZID=America/New_York19970902T090000', time => {
      expect(() => {
        parseDTStart(time);
      }).toThrowError(`Invalid "DTSTART" value "${time}"`);
    });

    test('TZID=America/New_York19970902T090000', time => {
      expect(() => {
        parseDTStart(time);
      }).toThrowError(`Invalid "DTSTART" value "${time}"`);
    });

    test('ap$', time => {
      expect(() => {
        parseDTStart(time);
      }).toThrowError(`Invalid "DTSTART" value "${time}"`);
    });
  });
});

describe('parseFrequency()', () => {
  describe('VALID', () => {
    test('SECONDLY', text => {
      expect(parseFrequency(text)).toEqual(text);
    });
  });

  describe('INVALID', () => {
    test('SECONDL', text => {
      expect(() => parseFrequency(text)).toThrowError(
        `Invalid FREQ value "${text}"`,
      );
    });
  });
});

describe('parseUntil()', () => {
  describe('VALID', () => {
    test('19970902T090000Z', text => {
      expect(
        parseUntil(
          text,
          StandardDateAdapter,
          new StandardDateAdapter(new Date()),
        ).toISOString(),
      ).toBe(
        new StandardDateAdapter(
          new Date(Date.UTC(1997, 8, 2, 9)),
        ).toISOString(),
      );
    });

    test('19970902T090000', text => {
      expect(
        parseUntil(
          text,
          StandardDateAdapter,
          new StandardDateAdapter(new Date()),
        ).toISOString(),
      ).toBe(new StandardDateAdapter(new Date(1997, 8, 2, 9)).toISOString());
    });
  });

  describe('INVALID', () => {
    test('19970902090000', text => {
      expect(() =>
        parseUntil(
          text,
          StandardDateAdapter,
          new StandardDateAdapter(new Date()),
        ),
      ).toThrowError(`Invalid ICAL date/time string "${text}"`);
    });
  });
});

describe('parseCount()', () => {
  describe('VALID', () => {
    test('3', text => {
      expect(parseCount(text)).toBe(3);
    });
  });

  describe('INVALID', () => {
    test('SECONDL', text => {
      expect(() => parseCount(text)).toThrowError(
        `Invalid COUNT value "${text}"`,
      );
    });
  });
});

describe('parseInterval()', () => {
  describe('VALID', () => {
    test('3', text => {
      expect(parseInterval(text)).toBe(3);
    });
  });

  describe('INVALID', () => {
    test('SECONDL', text => {
      expect(() => parseInterval(text)).toThrowError(
        `Invalid INTERVAL value "${text}"`,
      );
    });
  });
});

describe('parseBySecond()', () => {
  describe('VALID', () => {
    test('3', text => {
      expect(parseBySecond(text)).toEqual([3]);
    });

    test('3,4,6,7', text => {
      expect(parseBySecond(text)).toEqual([3, 4, 6, 7]);
    });
  });

  describe('INVALID', () => {
    test('SECONDL', text => {
      expect(() => parseBySecond(text)).toThrowError(
        `Invalid BYSECOND value "${text}"`,
      );
    });

    test('1,a,4,5', text => {
      expect(() => parseBySecond(text)).toThrowError(
        `Invalid BYSECOND value "a"`,
      );
    });
  });
});

describe('parseByMinute()', () => {
  describe('VALID', () => {
    test('3', text => {
      expect(parseByMinute(text)).toEqual([3]);
    });

    test('3,4,6,7', text => {
      expect(parseByMinute(text)).toEqual([3, 4, 6, 7]);
    });
  });

  describe('INVALID', () => {
    test('SECONDL', text => {
      expect(() => parseByMinute(text)).toThrowError(
        `Invalid BYMINUTE value "${text}"`,
      );
    });

    test('1,a,4,5', text => {
      expect(() => parseByMinute(text)).toThrowError(
        `Invalid BYMINUTE value "a"`,
      );
    });
  });
});

describe('parseByHour()', () => {
  describe('VALID', () => {
    test('3', text => {
      expect(parseByHour(text)).toEqual([3]);
    });

    test('3,4,60,7', text => {
      expect(parseByHour(text)).toEqual([3, 4, 60, 7]);
    });
  });

  describe('INVALID', () => {
    test('SECONDL', text => {
      expect(() => parseByHour(text)).toThrowError(
        `Invalid BYHOUR value "${text}"`,
      );
    });

    test('1,3,4,a', text => {
      expect(() => parseByHour(text)).toThrowError(`Invalid BYHOUR value "a"`);
    });
  });
});

describe('parseByDay()', () => {
  describe('VALID', () => {
    test('TU', text => {
      expect(parseByDay(text)).toEqual(['TU']);
    });

    test('1TU,WE,-3FR', text => {
      expect(parseByDay(text)).toEqual([['TU', 1], 'WE', ['FR', -3]]);
    });
  });

  describe('INVALID', () => {
    test('SECONDL', text => {
      expect(() => parseByDay(text)).toThrowError(
        `Invalid BYDAY value "${text}"`,
      );
    });

    test('1,3,4,a', text => {
      expect(() => parseByDay(text)).toThrowError(`Invalid BYDAY value "1"`);
    });

    test('+1TU,WE,-3FR', text => {
      expect(() => parseByDay(text)).toThrowError(`Invalid BYDAY value "+1TU"`);
    });
  });
});

describe('parseByMonthDay()', () => {
  describe('VALID', () => {
    test('3', text => {
      expect(parseByMonthDay(text)).toEqual([3]);
    });

    test('3,4,60,7', text => {
      expect(parseByMonthDay(text)).toEqual([3, 4, 60, 7]);
    });
  });

  describe('INVALID', () => {
    test('SECONDL', text => {
      expect(() => parseByMonthDay(text)).toThrowError(
        `Invalid BYMONTHDAY value "${text}"`,
      );
    });

    test('1,3,4,a', text => {
      expect(() => parseByMonthDay(text)).toThrowError(
        `Invalid BYMONTHDAY value "a"`,
      );
    });
  });
});

describe('parseByMonth()', () => {
  describe('VALID', () => {
    test('3', text => {
      expect(parseByMonth(text)).toEqual([3]);
    });

    test('3,4,60,7', text => {
      expect(parseByMonth(text)).toEqual([3, 4, 60, 7]);
    });
  });

  describe('INVALID', () => {
    test('SECONDL', text => {
      expect(() => parseByMonth(text)).toThrowError(
        `Invalid BYMONTH value "${text}"`,
      );
    });

    test('1,3,4,a', text => {
      expect(() => parseByMonth(text)).toThrowError(
        `Invalid BYMONTH value "a"`,
      );
    });
  });
});

describe('parseWkst()', () => {
  describe('VALID', () => {
    test('TU', text => {
      expect(parseWkst(text)).toBe('TU');
    });

    test('WE', text => {
      expect(parseWkst(text)).toBe('WE');
    });
  });

  describe('INVALID', () => {
    test('SECONDL', text => {
      expect(() => parseWkst(text)).toThrowError(
        `Invalid WKST value "${text}"`,
      );
    });

    test('SU,FR', text => {
      expect(() => parseWkst(text)).toThrowError(
        `Invalid WKST value "${text}"`,
      );
    });
  });
});

describe('parseICalStrings()', () => {
  describe('VALID', () => {
    test('DTSTART:19970902T090000\nRRULE:FREQ=DAILY;COUNT=10', ical => {
      expect(parseICalStrings([ical], StandardDateAdapter)).toEqual({
        rdates: [],
        exdates: [],
        exrules: [],
        rrules: [
          {
            start: new StandardDateAdapter(datetime(1997, 9, 2, 9)),
            frequency: 'DAILY',
            count: 10,
          },
        ],
      });
    });

    test(
      [
        'DTSTART:19970902T090000',
        'RRULE:FREQ=YEARLY;UNTIL=20000131T140000;BYMONTH=1;BYDAY=SU,MO,TU,WE,TH,FR,SA',
      ].join('\n'),
      ical => {
        expect(parseICalStrings([ical], StandardDateAdapter)).toEqual({
          rdates: [],
          exdates: [],
          exrules: [],
          rrules: [
            {
              start: new StandardDateAdapter(datetime(1997, 9, 2, 9)),
              frequency: 'YEARLY',
              until: new StandardDateAdapter(datetime(2000, 1, 31, 14)),
              byMonthOfYear: [1],
              byDayOfWeek: ['SU', 'MO', 'TU', 'WE', 'TH', 'FR', 'SA'],
            },
          ],
        });
      },
    );

    test(
      [
        'DTSTART:19970902T090000',
        'RRULE:FREQ=WEEKLY;UNTIL=19971007T000000Z;WKST=SU;BYDAY=TU,TH',
        'RDATE:19970714T123000Z',
        'RDATE:19970714T123000Z',
        'EXDATE:19970714T123000Z',
      ].join('\n'),
      ical => {
        expect(parseICalStrings([ical], StandardDateAdapter)).toEqual({
          rdates: [
            new StandardDateAdapter(new Date(Date.UTC(1997, 6, 14, 12, 30)), {
              timezone: 'UTC',
            }),
            new StandardDateAdapter(new Date(Date.UTC(1997, 6, 14, 12, 30)), {
              timezone: 'UTC',
            }),
          ],
          exdates: [
            new StandardDateAdapter(new Date(Date.UTC(1997, 6, 14, 12, 30)), {
              timezone: 'UTC',
            }),
          ],
          exrules: [],
          rrules: [
            {
              start: new StandardDateAdapter(datetime(1997, 9, 2, 9)),
              frequency: 'WEEKLY',
              until: new StandardDateAdapter(new Date(1997, 9, 6, 17)),
              weekStart: 'SU',
              byDayOfWeek: ['TU', 'TH'],
            },
          ],
        });
      },
    );

    test(
      [
        'DTSTART:19970907T090000',
        'RRULE:FREQ=MONTHLY;INTERVAL=2;COUNT=10;BYDAY=1SU,-1SU',
      ].join('\n'),
      ical => {
        expect(parseICalStrings([ical], StandardDateAdapter)).toEqual({
          rdates: [],
          exdates: [],
          exrules: [],
          rrules: [
            {
              start: new StandardDateAdapter(datetime(1997, 9, 7, 9)),
              frequency: 'MONTHLY',
              interval: 2,
              count: 10,
              byDayOfWeek: [['SU', 1], ['SU', -1]],
            },
          ],
        });
      },
    );

    test(
      [
        'DTSTART:19970910T090000',
        'RRULE:FREQ=SECONDLY;INTERVAL=18;COUNT=10;BYMONTHDAY=10,11,12',
        'RDATE:19970714T083000',
      ].join('\n'),
      ical => {
        expect(parseICalStrings([ical], StandardDateAdapter)).toEqual({
          rdates: [new StandardDateAdapter(datetime(1997, 7, 14, 8, 30))],
          exdates: [],
          exrules: [],
          rrules: [
            {
              start: new StandardDateAdapter(datetime(1997, 9, 10, 9)),
              frequency: 'SECONDLY',
              interval: 18,
              count: 10,
              byDayOfMonth: [10, 11, 12],
            },
          ],
        });
      },
    );
  });

  describe('INVALID', () => {
    test(
      [
        // wont parse timezone if given DateAdapter doesn't support timezones
        'DTSTART;TZID=America/New_York:19970902T090000',
        'RRULE:FREQ=YEARLY;UNTIL=20000131T140000Z;BYMONTH=1;BYDAY=SU,MO,TU,WE,TH,FR,SA',
      ].join('\n'),
      ical => {
        expect(() => {
          parseICalStrings([ical], StandardDateAdapter);
        }).toThrowError(
          'The `StandardDateAdapter` only supports datetimes in UTC or LOCAL time. ' +
            'You attempted to parse an ICAL string with a "America/New_York" timezone.',
        );
      },
    );

    test(
      [
        'DTSTART:19970101T090000',
        'RRULE:FREQ=YEARLY;INTERVAL=3;COUNT=10;BYYEARDAY=1,100,200',
      ].join('\n'),
      ical => {
        expect(() => {
          parseICalStrings([ical], StandardDateAdapter);
        }).toThrowError(
          '"BYYEARDAY" ' +
            'rule option is unsupported by rSchedule ' +
            'and I have no plans to implement it. Pull requests are welcome though.',
        );
      },
    );
  });
});

describe('ruleOptionsToIcalString()', () => {
  describe('VALID', () => {
    test('DTSTART:19970902T090000\nRRULE:FREQ=DAILY;COUNT=10', ical => {
      expect(
        ruleOptionsToIcalString(
          StandardDateAdapter,
          {
            start: new StandardDateAdapter(datetime(1997, 9, 2, 9)),
            frequency: 'DAILY',
            count: 10,
          },
          'RRULE',
        ),
      ).toBe(ical);
    });

    test(
      [
        'DTSTART:19970902T090000',
        'RRULE:FREQ=YEARLY;UNTIL=20000131T140000;BYMONTH=1;BYDAY=SU,MO,TU,WE,TH,FR,SA',
      ].join('\n'),
      ical => {
        expect(
          ruleOptionsToIcalString(
            StandardDateAdapter,
            {
              start: new StandardDateAdapter(datetime(1997, 9, 2, 9)),
              frequency: 'YEARLY',
              until: new StandardDateAdapter(datetime(2000, 1, 31, 14)),
              byMonthOfYear: [1],
              byDayOfWeek: ['SU', 'MO', 'TU', 'WE', 'TH', 'FR', 'SA'],
            },
            'RRULE',
          ),
        ).toBe(ical);
      },
    );

    test(
      [
        'DTSTART:19970902T090000',
        'RRULE:FREQ=WEEKLY;UNTIL=19971007T000000;WKST=SU;BYDAY=TU,TH',
      ].join('\n'),
      ical => {
        expect(
          ruleOptionsToIcalString(
            StandardDateAdapter,
            {
              start: new StandardDateAdapter(datetime(1997, 9, 2, 9)),
              frequency: 'WEEKLY',
              until: new StandardDateAdapter(datetime(1997, 10, 7)),
              weekStart: 'SU',
              byDayOfWeek: ['TU', 'TH'],
            },
            'RRULE',
          ),
        ).toBe(ical);
      },
    );

    test(
      [
        'DTSTART:19970907T090000',
        'RRULE:FREQ=MONTHLY;INTERVAL=2;COUNT=10;BYDAY=1SU,-1SU',
      ].join('\n'),
      ical => {
        expect(
          ruleOptionsToIcalString(
            StandardDateAdapter,
            {
              start: new StandardDateAdapter(datetime(1997, 9, 7, 9)),
              frequency: 'MONTHLY',
              interval: 2,
              count: 10,
              byDayOfWeek: [['SU', 1], ['SU', -1]],
            },
            'RRULE',
          ),
        ).toBe(ical);
      },
    );

    test(
      [
        'DTSTART:19970910T090000',
        'RRULE:FREQ=SECONDLY;INTERVAL=18;COUNT=10;BYMONTHDAY=10,11,12',
      ].join('\n'),
      ical => {
        expect(
          ruleOptionsToIcalString(
            StandardDateAdapter,
            {
              start: new StandardDateAdapter(datetime(1997, 9, 10, 9)),
              frequency: 'SECONDLY',
              interval: 18,
              count: 10,
              byDayOfMonth: [10, 11, 12],
            },
            'RRULE',
          ),
        ).toBe(ical);
      },
    );
  });
});


jest.config.js
/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */

module.exports = {
  // testRegex: '(/__tests__/.*|\\.(test|spec))\\.(ts|tsx|js)$',
  moduleDirectories: ['node_modules'],
  moduleFileExtensions: ['js', 'ts', 'tsx'],
  modulePathIgnorePatterns: ['tmp'],
  coveragePathIgnorePatterns: ['/node_modules/', '/tests/'],
  coverageThreshold: {
    global: {
      branches: 60,
      functions: 70,
      lines: 70,
      statements: 70,
    },
  },
  // collect coverage throws off the ts-jest source maps :(
  // see https://intellij-support.jetbrains.com/hc/en-us/community/posts/360004708619-TypeScript-and-Jest-debugger-stops-only-on-breakpoints-in-tests-never-in-source-files?page=1#community_comment_360000714019
  collectCoverage: false,
  verbose: false,
  testURL: 'http://localhost/',
  moduleNameMapper: {
    // more specific path maps need to come before less specific path maps
    // as `@rschedule/core` also matches `@rschedule/core/generators`
    '@local-tests/(.*)': '<rootDir>/../../tests/$1',
    '@local-tasks/(.*)': '<rootDir>/../../tasks/$1',
    '@rschedule/core/(.*)': '<rootDir>/../core/src/$1',
    '@rschedule/core': '<rootDir>/../core/src',
    '@rschedule/ical-tools/(.*)': '<rootDir>/../ical-tools/src/$1',
    '@rschedule/ical-tools': '<rootDir>/../ical-tools/src',
    '@rschedule/json-tools/(.*)': '<rootDir>/../json-tools/src/$1',
    '@rschedule/json-tools': '<rootDir>/../json-tools/src',
    '@rschedule/rule-tools/(.*)': '<rootDir>/../rule-tools/src/$1',
    '@rschedule/rule-tools': '<rootDir>/../rule-tools/src',
    '@rschedule/recurrence-rules-resolver/(.*)': '<rootDir>/../recurrence-rules-resolver/src/$1',
    '@rschedule/recurrence-rules-resolver': '<rootDir>/../recurrence-rules-resolver/src',
    '@rschedule/dayjs-date-adapter/(.*)': '<rootDir>/../dayjs-date-adapter/src/$1',
    '@rschedule/dayjs-date-adapter': '<rootDir>/../dayjs-date-adapter/src',
    '@rschedule/joda-date-adapter/(.*)': '<rootDir>/../joda-date-adapter/src/$1',
    '@rschedule/joda-date-adapter': '<rootDir>/../joda-date-adapter/src',
    '@rschedule/moment-date-adapter/(.*)': '<rootDir>/../moment-date-adapter/src/$1',
    '@rschedule/moment-date-adapter': '<rootDir>/../moment-date-adapter/src',
    '@rschedule/moment-tz-date-adapter/(.*)': '<rootDir>/../moment-tz-date-adapter/src/$1',
    '@rschedule/moment-tz-date-adapter': '<rootDir>/../moment-tz-date-adapter/src',
    '@rschedule/luxon-date-adapter/(.*)': '<rootDir>/../luxon-date-adapter/src/$1',
    '@rschedule/luxon-date-adapter': '<rootDir>/../luxon-date-adapter/src',
    '@rschedule/standard-date-adapter/(.*)': '<rootDir>/../standard-date-adapter/src/$1',
    '@rschedule/standard-date-adapter': '<rootDir>/../standard-date-adapter/src',
  },
  globals: {
    'ts-jest': {
      tsconfig: 'tsconfig.jest.json',
    },
  },
  preset: 'ts-jest',
  testMatch: ['**/*.test.ts'],
  testEnvironment: 'node',
};
